File: web/src/main/java/no/nav/modiapersonoversikt/config/endpoint/SfHenvendelseConfig.java
package no.nav.modiapersonoversikt.config.endpoint;

import no.nav.common.sts.SystemUserTokenProvider;
import no.nav.modiapersonoversikt.infrastructure.ping.ConsumerPingable;
import no.nav.modiapersonoversikt.infrastructure.types.Pingable;
import no.nav.modiapersonoversikt.service.ansattservice.AnsattService;
import no.nav.modiapersonoversikt.legacy.api.service.pdl.PdlOppslagService;
import no.nav.modiapersonoversikt.service.arbeidsfordeling.ArbeidsfordelingService;
import no.nav.modiapersonoversikt.service.sfhenvendelse.SfHenvendelseService;
import no.nav.modiapersonoversikt.service.sfhenvendelse.SfHenvendelseServiceImpl;
import org.springframework.context.annotation.Bean;

public class SfHenvendelseConfig {
    @Bean
    public SfHenvendelseService sfHenvendelseApi(
            PdlOppslagService pdlOppslagService,
            ArbeidsfordelingService arbeidsfordelingService,
            AnsattService ansattService,
            SystemUserTokenProvider systemUserTokenProvider
    ) {
        return new SfHenvendelseServiceImpl(
                pdlOppslagService,
                arbeidsfordelingService,
                ansattService,
                systemUserTokenProvider
        );
    }

    @Bean
    public Pingable sfHenvendelseApiPing(SfHenvendelseService service) {
        return new ConsumerPingable(
                "Salesforce - Henvendelse",
                service::ping
        );
    }
}

File: web/src/main/java/no/nav/modiapersonoversikt/service/sfhenvendelse/SfHenvendelseConfig.kt
package no.nav.modiapersonoversikt.service.sfhenvendelse

import no.nav.common.rest.client.RestClient
import no.nav.common.token_client.client.MachineToMachineTokenClient
import no.nav.common.token_client.client.OnBehalfOfTokenClient
import no.nav.modiapersonoversikt.config.interceptor.TjenestekallLoggingInterceptorFactory
import no.nav.modiapersonoversikt.consumer.norg.NorgApi
import no.nav.modiapersonoversikt.infrastructure.http.AuthorizationInterceptor
import no.nav.modiapersonoversikt.infrastructure.ping.ConsumerPingable
import no.nav.modiapersonoversikt.service.ansattservice.AnsattService
import no.nav.modiapersonoversikt.service.pdl.PdlOppslagService
import no.nav.modiapersonoversikt.service.sfhenvendelse.SfHenvendelseApiFactory.asTokenProvider
import no.nav.modiapersonoversikt.utils.bindTo
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import kotlin.time.Duration.Companion.seconds
import kotlin.time.toJavaDuration

@Configuration
open class SfHenvendelseConfig {
    @Bean
    open fun sfHenvendelseApi(
        pdlOppslagService: PdlOppslagService,
        norgApi: NorgApi,
        ansattService: AnsattService,
        oboTokenClient: OnBehalfOfTokenClient,
        machineToMachineTokenClient: MachineToMachineTokenClient,
        tjenestekallLoggingInterceptorFactory: TjenestekallLoggingInterceptorFactory,
    ): SfHenvendelseService {
        val oboTokenClient = oboTokenClient.bindTo(SfHenvendelseApiFactory.downstreamApi())
        val httpClient =
            RestClient
                .baseClient()
                .newBuilder()
                .addInterceptor(
                    tjenestekallLoggingInterceptorFactory("SF-Henvendelse") { request ->
                        requireNotNull(request.header("X-Correlation-ID")) {
                            "Kall uten \"X-Correlation-ID\" er ikke lov"
                        }
                    },
                ).addInterceptor(
                    AuthorizationInterceptor(oboTokenClient.asTokenProvider()),
                ).readTimeout(15.seconds.toJavaDuration())
                .build()

        return SfHenvendelseServiceImpl(
            pdlOppslagService = pdlOppslagService,
            norgApi = norgApi,
            ansattService = ansattService,
            httpClient = httpClient,
        )
    }

    @Bean
    open fun sfHenvendelseApiPing(service: SfHenvendelseService): ConsumerPingable =
        ConsumerPingable(
            "Salesforce - Henvendelse",
            service::ping,
        )
}
